Modules/modFileEncoding.ACM

```diff
@@ -0,0 +1,257 @@
+  ReDim bytsUCS2(2 + lenUTF8 * 2)
+  
+  ' Write the BOM
+  bytsUCS2(0) = 255
+  bytsUCS2(1) = 254
+  b = 2
+  
+  For i = LBound(bytsUTF8) To UBound(bytsUTF8)
+    j = UTF8CP_to_UCS2CP(bytsUTF8, i, CP_UCS2)
+    If j = 0 Then Exit Function
+    CopyMemory VarPtr(bytsUCS2(b)), VarPtr(CP_UCS2(0)), 2
+    b = b + 2
+    i = i + j - 1
+  Next
+  
+  ReDim Preserve bytsUCS2(b - 1)
+  UTF8_to_UCS2 = True
+  
+End Function
+
+Private Function UTF8CP_to_UCS2CP(bytsUTF8() As Byte, P As Long, bytsUCS2() As Byte) As Long
+  ' This function takes a UTF8 byte array and converts the codepoint at location p into a UCS2 byte array (length 2)
+  ' In the event of an error this function returns zero, otherwise it returns the number
+  ' of bytes which were used to encode the UTF8 character
+  
+  Dim c As Long
+  
+  UTF8CP_to_UCS2CP = 0 ' intitally set to zero (fail)
+  
+  If P > UBound(bytsUTF8) Then Exit Function
+  If P < LBound(bytsUTF8) Then Exit Function
+  
+  ReDim bytsUCS2(1) ' always 2 bytes
+  
+  If (bytsUTF8(P) And 128) = 0 Then '128 = 1000 0000 'this byte is 0xxx xxxx
+    ' We have a single byte UTF8
+    bytsUCS2(0) = bytsUTF8(P)
+    bytsUCS2(1) = 0
+    UTF8CP_to_UCS2CP = 1
+  ElseIf (bytsUTF8(P) And 224) = 192 Then  '224 = 1110 0000, 192 = 1100 0000 'this byte is 110x xxxx
+    If UBound(bytsUTF8) < P + 1 Then Exit Function 'check we have enough bytes left
+    If (bytsUTF8(P + 1) And 192) <> 128 Then Exit Function '192 = 1100 0000, 128 = 1000 0000 2nd byte must be 10xx xxxx
+    ' We have a valid 2 byte UTF8
+    c = bit16ShiftR((bytsUTF8(P) And 31), -6) '31 = 0001 1111
+    c = c Or (bytsUTF8(P + 1) And 63) '63 = 0011 1111
+    bytsUCS2(0) = c And 255 '255 = 1111 1111  least significant byte
+    bytsUCS2(1) = bit16ShiftR(c, 8) ' most significant Byte
+    UTF8CP_to_UCS2CP = 2
+  ElseIf (bytsUTF8(P) And 240) = 224 Then '240 = 1111 0000, 224 = 1110 0000 'this byte is 1110 xxxx
+    If UBound(bytsUTF8) < P + 2 Then Exit Function 'check we have enough bytes left
+    If (bytsUTF8(P + 1) And 192) <> 128 Then Exit Function '2nd byte must be 10xx xxxx
+    If (bytsUTF8(P + 2) And 192) <> 128 Then Exit Function '3rd byte must be 10xx xxxx
+    ' We have a valid 3 byte UTF8
+    c = bit16ShiftR((bytsUTF8(P) And 15), -12) '15 = 0000 1111
+    c = c Or bit16ShiftR((bytsUTF8(P + 1) And 63), -6) '63 = 0011 1111
+    c = c Or (bytsUTF8(P + 2) And 63) '63 = 0011 1111
+    bytsUCS2(0) = c And 255 '255 = 1111 1111  least significant byte
+    bytsUCS2(1) = bit16ShiftR(c, 8) ' most significant Byte
+    UTF8CP_to_UCS2CP = 3
+  Else
+    ' we cannot deal with 4 byte UTF8 so we get an error
+    Exit Function
+  End If
+    
+End Function
+
+Private Function UCS2CP_to_UTF8CP(bytsUCS2() As Byte, bytsUTF8() As Byte) As Boolean
+  ' This function takes a 2 byte zero based array encoded in UCS2 and returns a 1 to 3 byte array encoded in UTF8
+  
+  Dim i As Long ' this is used to represent the character we have

```

Comment:
I would expect `i` to be used as a loop counter.  use `x`  or `lng` instead?

Reply:
Changed to lngCodePoint